### PR TITLE
[cuDNN v8] Conv2D Backward/Transpose port to cuDNN v8 API

### PR DESCRIPTION
**Pull request by eqy**
Just a copy-paste pattern match to see if things work initially. Many parts of the code (such as the run() lambda) can likely be refactored to reduce duplication.
[will tag issues/people later]

#58858

### CODE REVIEW

**Comment by eqy** [comment]
CC @zasdfgbnm

**Review by zasdfgbnm — COMMENTED** [review]
Is it possible to have less copy-paste and more code reuse?

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 1]
```diff
@@ -121,56 +115,125 @@ void raw_cudnn_convolution_forward_out(
...
+  if (search != engine_cache.end()) {
+    run_conv_cfg(handle, input, grad_output, grad_weight, search->second);
+    return;
+  }
+
+  cudnn_frontend::EngineConfigList filtered_configs;
+  get_configs(filtered_configs, handle, CUDNN_BACKEND_OPERATION_CONVOLUTION_BACKWARD_FILTER_DESCRIPTOR, input, grad_output, grad_weight, key, padding, stride, dilation, deterministic, allow_tf32);
+  try_filtered_configs(filtered_configs, key, handle, input, grad_output, grad_weight);
+}
+
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:236]: Look like there are still lots of copy-pasted code here. Could we do something like this:

```C++
// used for convolution and transposed convolution's forward, backward data, and backward filter
void run_single_conv_operation(
    cudnnBackendDescriptorType_t operation,
    const Tensor& y, const Tensor& x, const Tensor& w,
    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
    bool benchmark, bool deterministic, bool allow_tf32)
{ ... }

void raw_cudnn_convolution_forward_out(
    const Tensor& output, const Tensor& input, const Tensor& weight,
    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
    bool benchmark, bool deterministic, bool allow_tf32)
{
  run_single_conv_operation(CUDNN_BACKEND_OPERATION_CONVOLUTION_FORWARD_DESCRIPTOR, ...);
}

void raw_cudnn_convolution_backward_input_out(
    const at::Tensor& grad_input,
    const at::Tensor& grad_output,
    const at::Tensor& weight,
    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
    bool benchmark, bool deterministic, bool allow_tf32)
{
  run_single_conv_operation(CUDNN_BACKEND_OPERATION_CONVOLUTION_BACKWARD_DATA_DESCRIPTOR, ...);
}

void raw_cudnn_convolution_backward_weight_out(
    const Tensor& grad_weight, const Tensor& grad_output, const Tensor& input,
    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
    bool benchmark, bool deterministic, bool allow_tf32)
{
  run_single_conv_operation(CUDNN_BACKEND_OPERATION_CONVOLUTION_BACKWARD_FILTER_DESCRIPTOR, ...);
}
```

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 2]
```diff
@@ -86,33 +88,25 @@ std::unordered_map<CacheKey, cudnn_frontend::ManagedOpaqueDescriptor, ParamsHash
...
+void run_conv_cfg(cudnnHandle_t handle, const Tensor& x, const Tensor& y, const Tensor& w, cudnn_frontend::ManagedOpaqueDescriptor cfg) {
     auto plan = cudnn_frontend::ExecutionPlanBuilder()
         .setHandle(handle)
         .setEngineConfig(cfg)
         .build();
 
     auto workspace_size = plan.getWorkspaceSize();
-    auto workspace = at::empty({workspace_size}, input.options().dtype(kByte));
-    void *data_ptrs[] = {input.data_ptr(), output.data_ptr(), weight.data_ptr()};
+    // TODO: is always using 'x' options() ok? e.g., in backward data
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:106]: It should be OK

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 3]
```diff
@@ -86,33 +88,25 @@ std::unordered_map<CacheKey, cudnn_frontend::ManagedOpaqueDescriptor, ParamsHash
...
-  key.input_alignment = getAlignment(input);
-  key.output_alignment = getAlignment(output);
-  key.weight_alignment = getAlignment(weight);
+void get_cachekey(CacheKey& key, const cudnnBackendDescriptorType_t operation, const Tensor& convoutput, const Tensor& convinput1, const Tensor& convinput2, IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups, bool deterministic, bool allow_tf32) {
+   setConvolutionParams(&key.params, convinput1, convinput2, padding, stride, dilation, groups, deterministic, allow_tf32);
+   key.operation = operation;
+   key.input_alignment = getAlignment(convinput1);
+   key.output_alignment = getAlignment(convoutput);
+   key.weight_alignment = getAlignment(convinput2);
+}
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:97]: How about making variable names in `CacheKey` and `get_cachekey` consistent with `run_single_conv`?
`input`/`convinput1` --> `x`
`output`/`convoutput` --> `y`
`weight`/`convinput2` --> `w`

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 4]
```diff
@@ -121,56 +115,108 @@ void raw_cudnn_convolution_forward_out(
...
-    run(search->second);
-    return;
-  }
+}
 
-  auto op = cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_CONVOLUTION_FORWARD_DESCRIPTOR)
-      .setxDesc(getTensorDescriptor(input, 'x', key.input_alignment))
-      .setyDesc(getTensorDescriptor(output, 'y', key.output_alignment))
-      .setwDesc(getTensorDescriptor(weight, 'w', key.weight_alignment))
+void get_configs(cudnn_frontend::EngineConfigList& filtered_configs, cudnnHandle_t handle, cudnnBackendDescriptorType_t desc, const Tensor& x, const Tensor& y, const Tensor& w, CacheKey key, IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, bool deterministic, bool allow_tf32) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:120]: Is `get_configs_from_heuristics` a better name?

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 5]
```diff
@@ -121,56 +115,108 @@ void raw_cudnn_convolution_forward_out(
...
+                    .setOperation(desc)
                     .build();
-
   auto& engine_configs = heuristics.getEngineConfig(heuristics.getEngineConfigCount());
   auto& fallback_list = fallback.getFallbackList();
 
-  cudnn_frontend::EngineConfigList filtered_configs;
-  filterEngineConfigs(engine_configs, filtered_configs, deterministic, allow_tf32, input.scalar_type());
-  filterEngineConfigs(fallback_list, filtered_configs, deterministic, allow_tf32, input.scalar_type());
+  // TODO: Is this ok or do we need to change type deduction based on descriptor?
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:145]: It should be OK

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 6]
```diff
@@ -121,56 +115,108 @@ void raw_cudnn_convolution_forward_out(
...
+  IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
+  bool benchmark, bool deterministic, bool allow_tf32) {
+  TORCH_CHECK(!benchmark, "not supported yet");
+  if (operation == CUDNN_BACKEND_OPERATION_CONVOLUTION_FORWARD_DESCRIPTOR && y.numel() == 0) {
+    return;
+  } else if (operation == CUDNN_BACKEND_OPERATION_CONVOLUTION_BACKWARD_DATA_DESCRIPTOR && x.numel() == 0) {
+    return;
+  } else if (operation == CUDNN_BACKEND_OPERATION_CONVOLUTION_BACKWARD_FILTER_DESCRIPTOR && w.numel() == 0) {
+    return;
+  }
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:172]: Can we move this to `raw_cudnn_convolution_forward_out`, `raw_cudnn_convolution_backward_input_out` and `raw_cudnn_convolution_backward_weight_out `?

File: aten/src/ATen/native/cudnn/Conv_v8.cpp [hunk 7]
```diff
@@ -74,103 +74,145 @@ void filterEngineConfigs(
...
+
+  try_filtered_configs(filtered_configs, key, handle, x, y, w);
+}
+
+void raw_cudnn_convolution_forward_out(
+    const Tensor& output, const Tensor& input, const Tensor& weight,
+    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
+    bool benchmark, bool deterministic, bool allow_tf32)
+{
+  if (y.numel() > 0) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cudnn/Conv_v8.cpp:185]: Did you test your change? `y` is not defined here.

Comment (eqy) [reply aten/src/ATen/native/cudnn/Conv_v8.cpp]: wow, can't believe I missed that, thanks
I think I restarted my environment and forgot to reset the experimental flag so I was rebuilding `Conv_v7.cpp` over and over...

**Review by zasdfgbnm — APPROVED** [review]
LGTM now!

**Comment by zasdfgbnm** [comment]
BTW, could you help fixing https://github.com/pytorch/pytorch/pull/51390#discussion_r634599724 in this PR as well?

**Comment by eqy** [comment]
ping; are these CI failures real?

**Comment by ngimel** [comment]
CI failures are not real

**Pull request closed without merging**